### ISSUE TITLE
[Question:] Support unkown values for categorical fields?

### ISSUE THREAD

**Issue by manugarri** [jpmml/jpmml-evaluator#56]
Lets say I train a Spark RandomForestClassifier, the exported pmml has for a specific field, the values:

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<PMML xmlns="http://www.dmg.org/PMML-4_3" version="4.3">
	<Header>
		<Application/>
		<Timestamp>2017-04-01T22:21:31Z</Timestamp>
	</Header>
	<DataDictionary>
		<DataField name="country" optype="categorical" dataType="string">
			<Value value="ESP"/>
			<Value value="RUS"/>
			<Value value="AUT"/>
		</DataField>
        ...
```

If a evaluation request has the value for this field `"country":"ITA"` for example, how can I setup the field so it allows unknown values? It doesnt have to do anything, just ignore the field if the value is unknown.

**Comment by vruusmann** [comment]
Some options, in no particular order:

- Remove all `Value` child elements. According to the PMML specification, if a `DataField` element does not define any `Value` elements (ie. the valid value space is undefined), then all values are considered to be valid.
- Add missing `Value` child elements. Should be easy when dealing with enumeration-type categorical features (such as country codes), where the "universe" of valid values is fixed and/or known beforehand.
- Change the value of the corresponding [`MiningField@invalidValueTreatment` attribute](http://dmg.org/pmml/v4-3/MiningSchema.html#xsdType_INVALID-VALUE-TREATMENT-METHOD) from `returnInvalid` to `asIs`. Please note that when dealing with ensemble models such as the Random Forest, you need to perform this change in more than one place (first, at the top-level parent `MiningModel` element, and then for every member `TreeModel` element). Luckily enough, this operation is trivial to automate using the Visitor approach.

**Comment by manugarri** [comment]
@vruusmann Thanks!. For what is worth, im using the evaluator with Openscoring, is there any way that Openscoring handles this? Maybe I should post the same question there?

**Comment by vruusmann** [comment]
> For what is worth, im using the evaluator with Openscoring, is there any way that Openscoring handles this?

Make a decision between the above three options, and implement it as a custom Visitor class. Then, edit Openscoring's configuration file, and define `modelRegistry.visitorClasses` property:
```
modelRegistry {
  visitorClasses = {"com.mycompany.pmml.OverrideInvalidValueTreatmentVisitor"}
}
```

> Maybe I should post the same question there?

Would be closed as duplicate of this one.

**Comment by manugarri** [comment]
ok, thanks. I am not fluent in Java, will do my best though. where can I find the Openscoring config file?

**Comment by vruusmann** [comment]
Default configuration:
https://github.com/openscoring/openscoring/blob/master/openscoring-service/src/main/resources/reference.conf

In the Openscoring README.md file, there's a longer explanation/example about overriding it with a custom configuration file.

**Comment by vruusmann** [comment]
The Visitor class for option three would roughly look like this:
```
public class OverrideInvalidValueTreatmentVisitor extends org.jpmml.model.visitors.AbstractSimpleVisitor {

  @Override
  public VisitorAction visit(MiningField miningField){
    miningField.setInvalidValueTreatment(InvalidValueTreatmentMethod.AS_IS);
    return super.visit(miningField);
  }
}
```

**Comment by manugarri** [comment]
awesome, thanks one more time for your help @vruusmann :)

**Comment by manugarri** [comment]
@vruusmann  I have been to tackle this issue, as I believe that implementing this feature as a visitor would be of benefit to the project (see [this fork](https://github.com/WanderaOrg/openscoring/commit/9ee328a10bf662a6314553eb906472a0e2dc0712)).

I did implement a new java class (we maintain our own fork of openscoring, I added it under openscoring-common just to test it)  named `OverrideInvalidValueTreatmentVisitor.java` The file looks like this

```
package org.openscoring.common;

import org.jpmml.model.visitors.AbstractSimpleVisitor;
import org.dmg.pmml.MiningField; 
import org.dmg.pmml.VisitorAction;
import org.dmg.pmml.InvalidValueTreatmentMethod;
	
public abstract class OverrideInvalidValueTreatmentVisitor extends AbstractSimpleVisitor {

  @Override
  public VisitorAction visit(MiningField miningField){
    miningField.setInvalidValueTreatment(InvalidValueTreatmentMethod.AS_IS);
    return super.visit(miningField);
  }
}
```

And I added it as you mentioned in the openscoring conf as

```
application {
	// List of JAX-RS Component class names that must be registered
	componentClasses = [
		"org.openscoring.service.NetworkSecurityContextFilter"
	]
}

modelRegistry {
	// List of Visitor class names that must be applied to the PMML class model object
	//visitorClasses = []
	visitorClasses = [
		"org.jpmml.model.visitors.LocatorNullifier",
                "org.jpmml.model.visitors.DataDictionaryCleaner",  
                 "org.jpmml.model.visitors.MiningSchemaCleaner",
                 "org.openscoring.common.OverrideInvalidValueTreatmentVisitor" //allows null miningfields
	]

	// Should the uploaded PMML documents be validated against the PMML XML Schema Definition (XSD) resource or not?
	// Validation is a one time operation that takes place during model deployment. It does not affect model evaluation or undeployment in any way.
	validate = false
}

networkSecurityContextFilter {
	// List of trusted IP addresses. An empty list defaults to all local network IP addresses.
	// A client that originates from a trusted IP address (as indicated by the value of the CGI variable REMOTE_ADDR) is granted the "admin" role.
	trustedAddresses = []
}
```
However, when I run openscoring server executable and PUT a model it fails with:
 
```
ERROR [qtp1365202186-16] (ModelResource.java:192) - Failed to load PMML document
java.lang.InstantiationException
	at sun.reflect.InstantiationExceptionConstructorAccessorImpl.newInstance(InstantiationExceptionConstructorAccessorImpl.java:48)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at org.openscoring.service.ModelRegistry.load(ModelRegistry.java:119)
	at org.openscoring.service.ModelResource.doDeploy(ModelResource.java:190)
	at org.openscoring.service.ModelResource.deploy(ModelResource.java:168)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
	at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:215)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:258)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:748)
```

Not sure how to fix this, what do you think I should modify?

Thanks as usual.

**Comment by vruusmann** [comment]
@manugarri You cannot instantiate an abstract class.  Remove the "abstract" modifier from the `OverrideInvalidValueTreatmentVisitor` class declaration, and rebuild & redeploy.

In the future, this kind of Visitor class should be made part of the [JPMML-Model](https://github.com/jpmml/jpmml-model) library. This is the central library, which is included in all downstream libraries such as the JPMML-Evaluator library, and ultimately the Openscoring REST web service.

**Comment by manugarri** [comment]
@vruusmann removing abstract from OverrideInvalidValueTreatmentVisitor yields the error:

```
openscoring_fork/openscoring-common/src/main/java/org/openscoring/common/OverrideInvalidValueTreatmentVisitor.java:[8,8] org.openscoring.common.OverrideInvalidValueTreatmentVisitor is not abstract and does not override abstract method visit(org.dmg.pmml.PMMLObject) in org.jpmml.model.visitors.AbstractSimpleVisitor
```

**Comment by vruusmann** [comment]
@manugarri  Make it `org.jpmml.model.visitors.AbstractVisitor` (instead of `org.jpmml.model.visitors.AbstractSimpleVisitor`) then. I really don't have time to compile and unit test all my responses.

**Comment by manugarri** [comment]
@vruusmann thanks for that, I know how tiring it is to evaluate every question/issue.  

After extending AbstractVisitor instead of AbstractSimpleVisitor openscoring does not complain about missing categoricals, but evaluation still fails with

```ERROR [qtp1365202186-15] (ModelResource.java:558) - Failed to evaluate
org.jpmml.evaluator.TypeCheckException (at or around line 82): Expected org.jpmml.evaluator.HasProbability, but got null
	at org.jpmml.evaluator.TypeUtil.cast(TypeUtil.java:485)
	at org.jpmml.evaluator.mining.SegmentResult.getTargetValue(SegmentResult.java:92)
	at org.jpmml.evaluator.mining.MiningModelEvaluator.aggregateProbabilities(MiningModelEvaluator.java:717)
	at org.jpmml.evaluator.mining.MiningModelEvaluator.evaluateClassification(MiningModelEvaluator.java:265)
	at org.jpmml.evaluator.mining.MiningModelEvaluator.evaluate(MiningModelEvaluator.java:207)
	at org.jpmml.evaluator.mining.MiningModelEvaluator.evaluate(MiningModelEvaluator.java:185)
	at org.openscoring.service.ModelResource.evaluate(ModelResource.java:700)
	at org.openscoring.service.ModelResource.doEvaluate(ModelResource.java:544)
	at org.openscoring.service.ModelResource.evaluate(ModelResource.java:324)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:205)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
	at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:215)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:258)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:748)
```

Which seems to be related with the [segment sizes](https://github.com/jpmml/jpmml-evaluator/blob/master/pmml-evaluator/src/main/java/org/jpmml/evaluator/mining/MiningModelEvaluator.java#L717) on the actual model. 

I think this might not be the best approach. Ideally there would be a data imputation that was supportive of missing values at evaluation time (like SKlearn OneHotEncoder for example).  I could use [PMML missingValueReplacement](http://dmg.org/pmml/v4-0-1/MiningSchema.html) to impute missing categoricals with a specific value.  I guess i can do it on a field basis, that would mean that the visitor would be model specific unfortunately.
 Ideally this would be handled by spark itself then automatically converted to a valid pmml, I dont think this is a possibility.

I think that for now, the easiest approach is just to edit the xml pmml model, what do you think?

**Comment by manugarri** [comment]
hey @vruusmann , this is how I fixed this at the end. 

1. Define a null constant that plays well with spark  (`"NULL"` for example).
2. At traning time, make sure that all the categorical fields have at least a record with the null constant on it (for example, adding an extra row to the training spark dataframe).
3. Modify the pmml -spark model output to to add to each MiningField both `invalidValueTreatment=asMissing` and `missingValueReplacement", "NULL"`.
4. Now the model evaluates and imputes those missing categoricals with NULL.

Hope this helps someone else in the future. I wrote a [brief article](http://blog.manugarri.com/handle-missing-categoricals-with-pmml/) on my blog on the matter.

Thanks so much for your help one more time.
